File: app/src/main/java/com/myflightbook/android/ActSelectMake.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2019 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.LinkMovementMethod;
import android.util.Log;
import android.widget.EditText;
import android.widget.SimpleExpandableListAdapter;
import android.widget.TextView;

import com.myflightbook.android.WebServices.MFBSoap;
import com.myflightbook.android.WebServices.MakesandModelsSvc;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;

import Model.MFBConstants;
import Model.MFBUtil;
import Model.MakesandModels;

public class ActSelectMake extends FixedExpandableListActivity {

    private int m_ModelID;

    private static class GetMakesTask extends AsyncTask<Void, Void, MFBSoap> {
        private ProgressDialog m_pd = null;
        private final AsyncWeakContext<ActSelectMake> m_ctxt;

        GetMakesTask(Context c, ActSelectMake asm) {
            super();
            m_ctxt = new AsyncWeakContext<>(c, asm);
        }

        @Override
        protected MFBSoap doInBackground(Void... params) {
            MakesandModelsSvc mms = new MakesandModelsSvc();
            MakesandModels[] rgmm = mms.GetMakesAndModels(m_ctxt.getContext());
            if (rgmm != null)
                ActNewAircraft.AvailableMakesAndModels = rgmm;

            return mms;
        }

        protected void onPreExecute() {
            m_pd = MFBUtil.ShowProgress(m_ctxt.getContext(), m_ctxt.getContext().getString(R.string.prgMakes));
        }

        protected void onPostExecute(MFBSoap svc) {
            try {
                if (m_pd != null)
                    m_pd.dismiss();
            } catch (Exception e) {
                Log.e(MFBConstants.LOG_TAG, Log.getStackTraceString(e));
            }

            ActSelectMake asm = m_ctxt.getCallingActivity();
            Context c = m_ctxt.getContext();

            if (asm == null || c == null)
                return;

            if (ActNewAircraft.AvailableMakesAndModels == null || ActNewAircraft.AvailableMakesAndModels.length == 0) {
                MFBUtil.Alert(c, c.getString(R.string.txtError), c.getString(R.string.errCannotRetrieveMakes));
            }
            else
                asm.populateList();
        }
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.selectmake);

        TextView tvSearch = findViewById(R.id.txtSearchProp);
        tvSearch.setHint(R.string.hintSearchModels);
        tvSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                populateList();
            }

            @Override
            public void afterTextChanged(Editable editable) { }
        });

        SwipeRefreshLayout srl = findViewById(R.id.swiperefresh);
        srl.setOnRefreshListener(() -> {
            srl.setRefreshing(false);
            refresh();
        });

        // make the hint for creating make/model a hyperlink
        TextView txtHint = findViewById(R.id.txtAddMakesHint);
        txtHint.setText(getString(R.string.lblAddMakes));
        txtHint.setMovementMethod(LinkMovementMethod.getInstance());
    }

    private void refresh() {
        new GetMakesTask(this, this).execute();
    }

    public void onResume() {
        super.onResume();

        Intent i = getIntent();
        m_ModelID = i.getIntExtra(ActNewAircraft.MODELFORAIRCRAFT, -1);
        populateList();
    }

    private void populateList() {
        // This maps the headers to the individual sub-lists.
        HashMap<String, HashMap<String, String>> headers = new HashMap<>();
        HashMap<String, ArrayList<HashMap<String, String>>> childrenMaps = new HashMap<>();

        // Keep a list of the keys in order
        ArrayList<String> alKeys = new ArrayList<>();
        String szKeyLast = "";
        int expandedGroupIndex = -1;

        String szRestrict = ((EditText) findViewById(R.id.txtSearchProp)).getText().toString().toUpperCase(Locale.getDefault());
        String[] rgRestrictStrings = szRestrict.split("\\W");

        // slice and dice into headers/first names
        if (ActNewAircraft.AvailableMakesAndModels != null) // should never be non-null, but seems to happen occasionally
        {
            for (int i = 0; i < ActNewAircraft.AvailableMakesAndModels.length; i++) {
                MakesandModels mm = ActNewAircraft.AvailableMakesAndModels[i];

                // reject anything that doesn't match the restriction
                boolean fIsMatch = true;
                for (String sz : rgRestrictStrings) {
                    if (sz.length() > 0 && !mm.Description.toUpperCase(Locale.getDefault()).contains(sz)) {
                        fIsMatch = false;
                        break;
                    }
                }
                if (!fIsMatch)
                    continue;

                // get the manufacturer for this property as the grouping key
                String szKey = mm.getManufacturer();
                if (szKey.compareTo(szKeyLast) != 0) {
                    alKeys.add(szKey);
                    szKeyLast = szKey;
                }

                HashMap<String, String> hmGroups = (headers.containsKey(szKey) ? headers.get(szKey) : new HashMap<>());
                hmGroups.put("sectionName", szKey);
                headers.put(szKey, hmGroups);

                // Get the array-list for that key, creating it if necessary
                ArrayList<HashMap<String, String>> alProps;
                alProps = (childrenMaps.containsKey(szKey) ? childrenMaps.get(szKey) : new ArrayList<>());

                HashMap<String, String> hmProperty = new HashMap<>();
                hmProperty.put("Description", mm.Description);
                hmProperty.put("Position", String.format(Locale.getDefault(), "%d", i));
                alProps.add(hmProperty);

                childrenMaps.put(szKey, alProps);

                // if this is the selected item, then expand this group index
                if (mm.ModelId == m_ModelID)
                    expandedGroupIndex = alKeys.size() - 1;
            }
        }

        boolean[] m_rgExpandedGroups = new boolean[alKeys.size()];
        if (expandedGroupIndex > 0 && expandedGroupIndex < alKeys.size())
            m_rgExpandedGroups[expandedGroupIndex] = true;
        if (m_rgExpandedGroups.length <= 5)
            for (int i = 0; i < m_rgExpandedGroups.length; i++)
                m_rgExpandedGroups[i] = true;

        // put the above into arrayLists, but in the order that the keys were encountered.  .values() is an undefined order.
        ArrayList<HashMap<String, String>> headerList = new ArrayList<>();
        ArrayList<ArrayList<HashMap<String, String>>> childrenList = new ArrayList<>();
        for (String s : alKeys) {
            headerList.add(headers.get(s));
            childrenList.add(childrenMaps.get(s));
        }

        final SimpleExpandableListAdapter adapter = new SimpleExpandableListAdapter(
                this,
                headerList,
                R.layout.grouprow,
                new String[]{"sectionName"},
                new int[]{R.id.propertyGroup},
                childrenList,
                R.layout.makemodelitem,
                new String[]{"Description"},
                new int[]{R.id.txtDescription}
        );
        setListAdapter(adapter);

        for (int i = 0; i < m_rgExpandedGroups.length; i++)
            if (m_rgExpandedGroups[i])
            getExpandableListView().expandGroup(i);

        getExpandableListView().setOnChildClickListener((parent, v, groupPosition, childPosition, id) -> {
                @SuppressWarnings("unchecked")
                HashMap<String, String> hmProp = (HashMap<String, String>) adapter.getChild(groupPosition, childPosition);
                int position = Integer.parseInt(hmProp.get("Position"));
                Intent i = new Intent();
                i.putExtra(ActNewAircraft.MODELFORAIRCRAFT, position);
                ActSelectMake.this.setResult(Activity.RESULT_OK, i);
                ActSelectMake.this.finish();
                return false;
            });
    }
}

File: app/src/main/java/com/myflightbook/android/ActSelectMake.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2025 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.view.View
import android.widget.EditText
import android.widget.ExpandableListView
import android.widget.SimpleExpandableListAdapter
import android.widget.TextView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.myflightbook.android.webservices.MakesandModelsSvc
import kotlinx.coroutines.launch
import model.MFBUtil.alert
import model.MakesandModels
import java.util.*

class ActSelectMake : FixedExpandableListActivity() {
    private var mModelID = 0

    private fun getMakes() {
        val act = this as Activity
        lifecycleScope.launch {
            ActMFBForm.doAsync<MakesandModelsSvc, Array<MakesandModels>?>(
                act,
                MakesandModelsSvc(),
                getString(R.string.prgMakes),
                { s -> s.getMakesAndModels(act) },
                { _, result ->
                    if (result == null || result.isEmpty()) {
                        alert(
                            act,
                            getString(R.string.txtError),
                            getString(R.string.errCannotRetrieveMakes)
                        )
                    } else {
                        ActNewAircraft.AvailableMakesAndModels = result
                        populateList()
                    }
                }
            )
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.selectmake)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.layout_root)) { view, insets ->
            val statusBarHeight = insets.getInsets(WindowInsetsCompat.Type.statusBars()).top
            view.setPadding(0, statusBarHeight, 0, 0)
            insets
        }
        val tvSearch = findViewById<TextView>(R.id.txtSearchProp)
        tvSearch.setHint(R.string.hintSearchModels)
        tvSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                populateList()
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        val srl = findViewById<SwipeRefreshLayout>(R.id.swiperefresh)
        srl.setOnRefreshListener {
            srl.isRefreshing = false
            refresh()
        }

        // make the hint for creating make/model a hyperlink
        val txtHint = findViewById<TextView>(R.id.txtAddMakesHint)
        txtHint.text = getString(R.string.lblAddMakes)
        txtHint.movementMethod = LinkMovementMethod.getInstance()
    }

    private fun refresh() {
        getMakes()
    }

    public override fun onResume() {
        super.onResume()
        val i = intent
        mModelID = i.getIntExtra(ActNewAircraft.MODELFORAIRCRAFT, -1)
        populateList()
    }

    private fun populateList() {
        // This maps the headers to the individual sub-lists.
        val headers = HashMap<String, HashMap<String, String>>()
        val childrenMaps = HashMap<String, ArrayList<HashMap<String, String>>>()

        // Keep a list of the keys in order
        val alKeys = ArrayList<String>()
        var szKeyLast: String? = ""
        var expandedGroupIndex = -1
        val szRestrict =
            (findViewById<View>(R.id.txtSearchProp) as EditText).text.toString().uppercase(Locale.getDefault())
        val rgRestrictStrings = szRestrict.split("\\W".toRegex()).toTypedArray()

        val rgmm = ActNewAircraft.AvailableMakesAndModels

        // slice and dice into headers/first names
        if (rgmm != null) // should never be non-null, but seems to happen occasionally
        {
            for (i in 0 until rgmm.count()) {
                val mm = rgmm[i]
                // reject anything that doesn't match the restriction
                var fIsMatch = true
                for (sz in rgRestrictStrings) {
                    if (sz.isNotEmpty() && !mm.description.uppercase(Locale.getDefault())
                            .contains(sz)
                    ) {
                        fIsMatch = false
                        break
                    }
                }
                if (!fIsMatch) continue

                // get the manufacturer for this property as the grouping key
                val szKey = mm.manufacturer
                if (szKey.compareTo(szKeyLast!!) != 0) {
                    alKeys.add(szKey)
                    szKeyLast = szKey
                }
                val hmGroups: HashMap<String, String> =
                    (if (headers.containsKey(szKey)) headers[szKey]!! else HashMap())
                hmGroups["sectionName"] = szKey
                headers[szKey] = hmGroups

                // Get the array-list for that key, creating it if necessary
                val alProps: ArrayList<HashMap<String, String>> =
                    if (childrenMaps.containsKey(szKey)) childrenMaps[szKey]!! else ArrayList<HashMap<String, String>>()
                val hmProperty = HashMap<String, String>()
                hmProperty["Description"] = mm.description
                hmProperty["Position"] = String.format(Locale.getDefault(), "%d", i)
                alProps.add(hmProperty)
                childrenMaps[szKey] = alProps

                // if this is the selected item, then expand this group index
                if (mm.modelId == mModelID) expandedGroupIndex = alKeys.size - 1
            }
        }
        val mrgExpandedGroups = BooleanArray(alKeys.size)
        if (expandedGroupIndex > 0 && expandedGroupIndex < alKeys.size) mrgExpandedGroups[expandedGroupIndex] =
            true
        if (mrgExpandedGroups.size <= 5) Arrays.fill(mrgExpandedGroups, true)

        // put the above into arrayLists, but in the order that the keys were encountered.  .values() is an undefined order.
        val headerList = ArrayList<HashMap<String, String>>()
        val childrenList = ArrayList<ArrayList<HashMap<String, String>>>()
        for (s in alKeys) {
            headerList.add(headers[s]!!)
            childrenList.add(childrenMaps[s]!!)
        }
        val adapter = SimpleExpandableListAdapter(
            this,
            headerList,
            R.layout.grouprow, arrayOf("sectionName"), intArrayOf(R.id.propertyGroup),
            childrenList,
            R.layout.makemodelitem, arrayOf("Description"), intArrayOf(R.id.txtDescription)
        )
        setListAdapter(adapter)
        for (i in mrgExpandedGroups.indices) if (mrgExpandedGroups[i]) expandableListView!!.expandGroup(
            i
        )
        expandableListView!!.setOnChildClickListener { _: ExpandableListView?, _: View?, groupPosition: Int, childPosition: Int, _: Long ->
            @Suppress("UNCHECKED_CAST")
            val hmProp = adapter.getChild(groupPosition, childPosition) as HashMap<String, String>
            val position = hmProp["Position"]!!.toInt()
            val i = Intent()
            i.putExtra(ActNewAircraft.MODELFORAIRCRAFT, position)
            this@ActSelectMake.setResult(RESULT_OK, i)
            finish()
            false
        }
    }
}